x/auth/vesting/types/period.go

```diff
@@ -26,3 +27,257 @@ func (vp Periods) String() string {
 	return strings.TrimSpace(fmt.Sprintf(`Vesting Periods:
 		%s`, strings.Join(periodsListString, ", ")))
 }
+
+// A schedule is an increasing step function of Coins over time.
+// It's specified as an absolute start time and a sequence of relative
+// periods, with each steps at the end of a period. A schedule may also
+// give the time and total value at the last step, which can speed
+// evaluation of the step function after the last step.
+
+// ReadSchedule returns the value of a schedule at readTime.
+func ReadSchedule(startTime, endTime int64, periods []Period, totalCoins sdk.Coins, readTime int64) sdk.Coins {
+	coins := sdk.NewCoins()
+
+	if readTime <= startTime {
+		return coins
+	}
+	if readTime >= endTime {
+		return totalCoins
+	}
+
+	time := startTime
+
+	for _, period := range periods {
+		x := readTime - time
+		if x < period.Length {
+			break
+		}
+		coins = coins.Add(period.Amount...)
+		time += period.Length
+	}
+
+	return coins
+}
+
+// max64 returns the maximum of its inputs.
+func max64(i, j int64) int64 {
+	if i > j {
+		return i
+	}
+	return j
+}
+
+// min64 returns the minimum of its inputs.
+func min64(i, j int64) int64 {
+	if i < j {
+		return i
+	}
+	return j
+}
+
+func coinsMin(a, b sdk.Coins) sdk.Coins {
+	min := sdk.NewCoins()
+	for _, coinA := range a {
+		denom := coinA.Denom
+		bAmt := b.AmountOf(denom)
+		minAmt := coinA.Amount
+		if minAmt.GT(bAmt) {
+			minAmt = bAmt
+		}
+		if minAmt.IsPositive() {
+			min = min.Add(sdk.NewCoin(denom, minAmt))
+		}
+	}
+	return min
+}
+
+// DisjunctPeriods returns the union of two vesting period schedules.
+// The returned schedule is the union of the vesting events, with simultaneous
+// events combined into a single event.
+// Returns new start time, new end time, and merged vesting events, relative to
+// the new start time.
+func DisjunctPeriods(startP, startQ int64, p, q []Period) (int64, int64, []Period) {
```

Comment:
this code is really hard to read and thus not very maintainable. What is P, what is Q?

Reply:
Inputs `p` and `q` are the sequences of periods of the two schedules to merge. It's just notation, like inputs `a` and `b` to `coinsMin()` right above.

I've made it more consistent (but less compact) by renaming `p` and `q` to `periodsP` and `periodsQ` respectively.